File: concrete/src/main/java/pcb/uwu/amount/composite/fundamental/Hertz.java
package pcb.uwu.amount.composite.fundamental;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.fundamental.FrequencyUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.composite.fundamental.HertzUnit.HERTZ;

public class Hertz extends Frequency {
	
	// region constructors
	
	public Hertz(Number value) {
		super(value, HERTZ);
	}

	public Hertz(Number value, Magnitude magnitude) {
		super(value, magnitude, HERTZ);
	}

	public Hertz(String value) {
		super(value, HERTZ);
	}

	public Hertz(String value, Magnitude magnitude) {
		super(value, magnitude, HERTZ);
	}

	public Hertz(BigDecimal value) {
		super(value, HERTZ);
	}

	public Hertz(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, HERTZ);
	}

	public Hertz(BigDecimalAmount amount) {
		super(amount, HERTZ);
	}

	public Hertz(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, HERTZ);
	}
	
	// endregion

	// region implement UnitAmount

	@Override
	public Hertz plus(UnitAmount<FrequencyUnit> other, MathContext mathContext) {
		return new Hertz(plusAmount(other, mathContext));
	}

	@Override
	public Hertz minus(UnitAmount<FrequencyUnit> other, MathContext mathContext) {
		return new Hertz(minusAmount(other, mathContext));
	}

	@Override
	public Hertz multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Hertz(multipliedByScalar(other, mathContext));
	}

	@Override
	public Hertz dividedBy(BigDecimal other, MathContext mathContext) {
		return new Hertz(dividedByScalar(other, mathContext));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/derived/acoustics/Hertz.kt
package pcb.uwu.amount.derived.acoustics

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.acoustics.FrequencyUnit
import pcb.uwu.unit.derived.acoustics.HertzUnit

class Hertz : Frequency
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, HertzUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, HertzUnit)

    // region UnitAmount

    override fun plus(frequency: UnitAmount<FrequencyUnit>) =
        Hertz(this.amount + (frequency to HertzUnit).amount)

    override fun minus(frequency: UnitAmount<FrequencyUnit>) =
        Hertz(this.amount - (frequency to HertzUnit).amount)

    override fun times(scalar: Number) =
        Hertz(this.amount * scalar)

    override fun div(scalar: Number) =
        Hertz(this.amount / scalar)

    // endregion
}